File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    int[] dr = {-1, 0, 1, 0};
    int[] dc = {0, -1, 0, 1};
    
    public int solution(int[][] maps) {
        
        int ROW = maps.length;
        int COL = maps[0].length;
        
        //row, col, 이동 카운트
        Queue<List<Integer>> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.get(2)));
        pq.add(Arrays.asList(0, 0, 1));
        // boolean[][] visited = new boolean[ROW][COL];
        
        while (!pq.isEmpty()) {
            List<Integer> cur = pq.poll();
            int row = cur.get(0); 
            int col = cur.get(1);
            int count = cur.get(2);
            
            //최적의 해를 구했으므로 리턴
            if (row == ROW - 1 && col == COL - 1) {
                return count;
            }
            // visited[row][col] = true;

            for (int d = 0; d < 4; d++) {
                int nextRow = row + dr[d];
                int nextCol = col + dc[d];
                
                if (nextRow < 0 || ROW <= nextRow || nextCol < 0 || COL <= nextCol) {//맵 경계처리
                    continue;
                }
                if (maps[nextRow][nextCol] != 0) {
                    //이미 지나온 경로는 더 이상 거치지 않도록 0 설정
                    maps[nextRow][nextCol] = 0;
                    pq.add(Arrays.asList(nextRow, nextCol, count + 1));
                }
                
            }
        }
        // pq의 모든 것이 빠져도 도착 못하면 갈 수 있는 방법이 없으므로 -1 반환
        return -1; 
    }
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    
    const dr = [-1, 0, 1, 0];
    const dc = [0, -1, 0, 1];
    
    const ROW = maps.length;
    const COL = maps[0].length;
    
    let visited = Array(ROW).fill().map(e => Array(COL).fill(false));
    
    const bfs = ((r, c) => {
        let q = new Array();
        q.push([r, c]);
        visited[r][c] = true;
        let count = 1;
        
        while(q.length !== 0) {
            const size = q.length;
            for (let i = 0; i < size; i++) {
                let current = q.shift();
                // console.log(current);
                
                if (current[0] == ROW - 1 && current[1] == COL - 1) {
                    return count;
                }
                
                for (let d = 0; d < 4; d++) {
                    let nextRow = current[0] + dr[d];
                    let nextCol = current[1] + dc[d];

                    if (nextRow < 0 || ROW <= nextRow 
                        || nextCol < 0 || COL <= nextCol) continue;

                    if (!visited[nextRow][nextCol] && maps[nextRow][nextCol] == 1) {
                        q.push([nextRow, nextCol]);
                        visited[nextRow][nextCol] = true; 
                    }
                }
            }
            count++;
        }
        return -1;
    })
    
    const result = bfs(0, 0, visited, maps, dr, dc, ROW, COL);
    var answer = result;
    return answer;
}